app/src/androidTest/java/Cryptohopper/Android/SDK/hopper/HopperInstrumentedTest.kt

```diff
@@ -0,0 +1,461 @@
+package Cryptohopper.Android.SDK.exchange
+
+import Cryptohopper.Android.SDK.helper.Const
+import Cryptohopper.Android.SDK.helper.Const.API_KEY
+import Cryptohopper.Android.SDK.helper.Const.API_PASSWORD
+import Cryptohopper.Android.SDK.helper.Const.API_USER
+import androidx.test.platform.app.InstrumentationRegistry
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.github.mervick.aes_everywhere.Aes256
+import cryptohopper.android.sdk.CryptohopperAuth
+import cryptohopper.android.sdk.SharedModels.ConfigModels.HopperAPIEnvironment
+import kotlinx.coroutines.*
+import org.junit.Assert
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.Before
+
+@DelicateCoroutinesApi
+@RunWith(AndroidJUnit4::class)
+class HopperInstrumentedTest {
+
+    @Before
+    fun setup() {
+        val appContext = InstrumentationRegistry.getInstrumentation().targetContext
+        CryptoHopperConfig.configure(
+            appContext, API_KEY,
+            HopperAPIEnvironment.Production
+        )
+        callAuthenticationWithAccurateDetails()
+    }
+
+    private fun callAuthenticationWithAccurateDetails() {
+        val userAgent = Aes256.encrypt(API_USER, Const.API_AGENT)
+        CryptohopperAuth.login(
+            username = API_USER,
+            password = API_PASSWORD,
+            verificationCode = null,
+            userAgent = userAgent
+        ) { result, error ->
+            Assert.assertNull(error)
+            Assert.assertNotNull(result)
+        }
+    }
+
+    //******************* Positive cases ***************************
+
+    // ------------- GENERAL ----------------------------------------
+
+    @Test
+    fun when_the_given_getAllHoppers_Endpoint_is_called_with_correct_token_then_it_must_return_hopper_list() {
+        CryptohopperHopper.getAllHoppers(null, null, null) { hoppers, error ->
+            Assert.assertNull(error)
+            Assert.assertNotNull(hoppers)
+
+            hoppers?.forEach {
+                Assert.assertNotNull(it.id)
+                Assert.assertNotNull(it.name)
```

Comment:
There is no guarantee that some of the variables of the hopper is not null . So it may be null or even not send if its not available .